crates/ruma-api-macros/src/request/outgoing.rs

```diff
@@ -15,39 +15,41 @@ impl Request {
 
         let method = &self.method;
         let error_ty = &self.error_ty;
-        let request_path_string = if self.has_path_fields() {
-            let mut format_string = self.path.value();
-            let mut format_args = Vec::new();
 
-            while let Some(start_of_segment) = format_string.find(':') {
-                // ':' should only ever appear at the start of a segment
-                assert_eq!(&format_string[start_of_segment - 1..start_of_segment], "/");
+        let (unstable_path, r0_path, stable_path) = if self.has_path_fields() {
+            let mapper = |s: &LitStr| -> TokenStream {
+                let (s, a) = util::convert_path_string(s.value(), &percent_encoding);
 
-                let end_of_segment = match format_string[start_of_segment..].find('/') {
-                    Some(rel_pos) => start_of_segment + rel_pos,
-                    None => format_string.len(),
-                };
-
-                let path_var = Ident::new(
-                    &format_string[start_of_segment + 1..end_of_segment],
-                    Span::call_site(),
-                );
-                format_args.push(quote! {
-                    #percent_encoding::utf8_percent_encode(
-                        &::std::string::ToString::to_string(&self.#path_var),
-                        #percent_encoding::NON_ALPHANUMERIC,
-                    )
-                });
-                format_string.replace_range(start_of_segment..end_of_segment, "{}");
-            }
+                quote! {
+                    format_args!(#s, #(#a),*)
+                }
+            };
 
-            quote! {
-                format_args!(#format_string, #(#format_args),*)
-            }
+            (
+                self.unstable_path.as_ref().map(mapper),
+                self.r0_path.as_ref().map(mapper),
+                self.stable_path.as_ref().map(mapper),
+            )
         } else {
-            quote! { metadata.path.to_owned() }
+            (
+                self.unstable_path
+                    .as_ref()
+                    .map(|_| quote! { format_args!("{}", metadata.unstable_path.unwrap()) }),
```

Comment:
This should work too, right? That way we don't have unnecessary `.unwrap()`s in the generated code:

```suggestion
                    .map(|path| quote! { format_args!(#path) }),
```